### ISSUE TITLE
Cluster Upgrade

### ISSUE THREAD

**Issue by lavalamp** [GoogleCloudPlatform/kubernetes#2524]
A Kubernetes cluster is turned on, becomes important, and must not go down. But the versions of its components should increase with time. This issue collects the items that need to be solved for this to be possible.

Things that need upgrading:

1. Node components:
 a. Kubelet
 b. Proxy
 c. Docker
 d. Kernel/OS

2. Master components, current configuration:
  a. apiserver
  b. controller-manager
  c. scheduler
  d. etcd
  e. Kernel/OS

2. Master components, hypothetical all-in-one binary configuration:
 a. all-in-one binary
 b. etcd
 c. Kernel/OS

2. Master components, self-hosted (master components in pods) configuration:
 a. For each master component (scheduler, rep. ctrl., etc), a rolling upgrade that we preach for applications.
 b. Plus a regular node upgrade

3. Objects in storage, when storage version is changed.

Mechanisms for upgrading:

1. "Easy" node upgrade
 a. Have a node image, including kernel/OS/docker/kubelet/proxy. Bring up new node, delete old node.
 b. Requires SLA for pods.
 c. Have to do this for kernel upgrades anyway, so why not use it for every upgrade?

2. "In-place" binary upgrade - for kubelet, proxy, non-self-hosted master components
 a. Get the new binary
 b. Quit running the old one; start running the new one.
 c. If new one is bad, what then?
 d. Requires a baby sitter or something to go back to the old if the new doesn't start.
 e. But also have to be able to upgrade the babysitter...

3. Rolling upgrade - for self-hosted master components.
 a. Bring up new replicationController
 b. Verify that new components are healthy (and functional?)
 c. Turn down old replicationController.
 d. (for replicas > 1, do this in increment-new/decrement-old pairs)

4. Objects in storage, when storage version is changed.
 a. Loop over all objects in database, perform version upgrade.
 b. Probably OK if eventually consistent; to perform on a running cluster, first tell all components writing/updating to storage to start using the new version (i.e., read old or new version, write new version). Then just loop through all objects and if they have the wrong version, upgrade and write them. Our conversion code is pretty easy to call so this should be relatively painless.
 c. Same procedure works for downgrade, as long as our conversion code round-trips all objects.

Requirements for upgrading:

1. Must be able to rollback. Need policy for how many versions back we ought to be able to go.
5. All upgrades should proceed slowly enough that humans can react to a problem before the whole cluster is broken.
2. Storage scheme & api version must change independently from each other and from component version. (already true)
3. Upgrading the storage scheme is orthogonal from upgrading binaries.
 a. Upgrading should be non-destructive; old deleted fields should not be permanently deleted, perhaps could be stored as an annotation.
 b. Cluster admins would be well-advised not to do this without multiple versions to roll back to.
4. Downgrading storage scheme: decision required as to whether we should support this.
6. Rules for safe storage versioning: may never remove or modify a field from an existing supported version. May add a field. May not populate a field until there's an old version to roll back to that can deal with the field being set (not important as long as encoding/json ignores unknown fields).

Paths that *must* be tested:

1. e2e tests, for each of the last N releases.
 a. Make and populate a cluster. Upgrade to latest nightly build. Verify functionality. Downgrade to original release. Verify functionality.
2. For each storage version change, from each of the last N storage versions (v1beta1, etc):
 a. Use fuzzing to generate a bunch of objects; test roll forward and roll backwards.

Nice-to-have tests:

1. Test etcd upgrade & rollback between two etcd versions that are known to both work.
2. Test upgrade/rollback of individual cluster components. (I.e., simulate emergency fix for kubelet.)

Research needed:

1. How to trigger a docker-daemon upgrade & rollback. (If we decide to do the "in-place" binary upgrades.)
2. How to trigger an etcd upgrade & rollback.

Decisions needed:

1. Implement only "Easy" node upgrade, or both "easy" & "in-place" node upgrades?
2. Can't block this on self-hosting master components. But perhaps work on them in parallel.
3. The above determine whether we need to write a mechanism for the "in-place" binary upgrade.

Let me know if I missed something and I'll edit.

**Comment by jasonkuhrt** [comment]
@brendanburns In regards to the instructions here https://github.com/GoogleCloudPlatform/kubernetes/issues/3315#issuecomment-69223228 would I need to do all this from any `k8s` node in the cluster, the master, or elsewhere (local machine etc.). Pretty sure it would be the master but just want to make sure

**Comment by a-robinson** [comment]
You should be able to run kube-push from anywhere (including your local machine) as long as your local kubecfg/kubectl is set up to talk to the cluster that you want to try upgrading.

**Comment by jasonkuhrt** [comment]
@a-robinson It is except I have to pass

```
--insecure-skip-tls-verify=true --server=x.x.x.x
```

Every time. How does that factor in here?

**Comment by a-robinson** [comment]
The master certs for GKE clusters are typically managed by gcloud. Your options are to either continue passing the skip-tls flag or to grab them from where gcloud put them and feed them into the command.

cc @jlowdermilk for details on the cert managed by gcloud.

**Comment by jlowdermilk** [comment]
If you have used gcloud to create or interact with the GKE cluster, gcloud will have cached credentials for the master. You should be able to see them by running
```
$ ls ~/.config/gcloud/kubernetes/{project-name}.{zone}.{cluster-name}/
ca.crt  cluster.json  kubecfg.crt  kubecfg.key  kubernetes_auth
```
You can copy those out and/or manually specify the paths via the appropriate kubectl/kubecfg flags.

**Comment by jasonkuhrt** [comment]
Hm, I tried but the cluster master is being looked for in the wrong zone:

```
$ cluster/kube-push.sh

Updating cluster using provider: gce
Project: littlebits-electronics
Project: littlebits-electronics

ERROR: (gcloud.compute.instances.describe) Could not fetch resource:
 - The resource 'projects/littlebits-electronics/zones/us-central1-b/instances/k8s-cluster-test1-master' was not found
```

I need it to use `us-central1-a`.

**Comment by jlowdermilk** [comment]
Edit https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/config-default.sh

**Comment by jasonkuhrt** [comment]
@jlowdermilk Thanks.

**Comment by jasonkuhrt** [comment]
Hm, now this error:

```
ERROR: (gcloud.compute.instances.describe) Could not fetch resource:
 - The resource 'projects/littlebits-electronics/zones/us-central1-a/instances/k8s-cluster-test1-minion-1' was not found

Did not find k8s-cluster-test1-minion-1
ERROR: (gcloud.compute.instances.describe) Could not fetch resource:
 - The resource 'projects/littlebits-electronics/zones/us-central1-a/instances/k8s-cluster-test1-minion-2' was not found

Did not find k8s-cluster-test1-minion-2
ERROR: (gcloud.compute.instances.describe) Could not fetch resource:
 - The resource 'projects/littlebits-electronics/zones/us-central1-a/instances/k8s-cluster-test1-minion-3' was not found

Did not find k8s-cluster-test1-minion-3
ERROR: (gcloud.compute.instances.describe) Could not fetch resource:
 - The resource 'projects/littlebits-electronics/zones/us-central1-a/instances/k8s-cluster-test1-minion-4' was not found

Did not find k8s-cluster-test1-minion-4
Could not detect Kubernetes minion nodes.  Make sure you've launched a cluster with 'kube-up.sh'
```

I will try editing this line https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/config-default.sh#L36 to be `node` but can someone explain to me what's going on. `minion` is the new or old terminology? I assume new...

**Comment by jlowdermilk** [comment]
minion is old terminology. We're still in the process of updating names/references in code and examples #1111.

**Comment by jasonkuhrt** [comment]
Now a new issue:

```
Project: littlebits-electronics
Project: littlebits-electronics
ERROR: (gcloud.compute.instances.describe) Could not fetch resource:
 - The resource 'projects/littlebits-electronics/zones/us-central1-a/instances/k8s-cluster-test1-node-4' was not found

Did not find k8s-cluster-test1-node-4
Running: cluster/../cluster/gce/../../cluster/../cluster/gce/../../cluster/../cluster/gce/../../platforms/darwin/amd64/kubectl get minions -o template -t {{range.items}}{{.id}}
{{end}}
F0108 16:10:58.410829   72413 get.go:75] Get https://146.148.60.38/api/v1beta1/minions?namespace=default: x509: certificate signed by unknown authority
```

Not sure why it thinks I have `4` nodes when I actually have `3`. After that there is a cert error but not sure it matters, is related, is unrelated, or what.

**Comment by jasonkuhrt** [comment]
Ahhh... after the failed updates it seems that my `gcloud` commands are failing on cert issues now??

```
$ gcloud preview container pods list

ERROR: F0108 16:31:10.279838   73593 get.go:74] Get https://146.148.60.38/api/v1beta1/pods?labels=&namespace=default: x509: certificate signed by unknown authority
```

**Comment by roberthbailey** [comment]
It will think you have 4 nodes unless you change 
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/config-default.sh#L23 to match the actual size of your cluster.

**Comment by jasonkuhrt** [comment]
Sorry I'm really not looking at this file closely enough. Thanks @roberthbailey. Stills stumped about the security issues, digging around still...

**Comment by jlowdermilk** [comment]
If the update succeeded, at least enough that the master was updated, it will generate new cert files. gcloud doesn't know that, and will continue using cached ones. You can force a refetch by running a gcloud command with --purge-config-cache. e.g.
```
gcloud preview container kubectl --purge-config-cache get nodes
```

**Comment by lavalamp** [comment]
Hey guys... sorry to disagree with @brendandburns, but could you take this discussion back to #3315? The fact that you're having this conversation is evidence that this issue hasn't been fixed yet, but this issue is trying to solve the problem at a higher level of abstraction.

**Comment by pires** [comment]
In GCE, you could snapshot&clone current cluster VMs, upgrade everything, test if everything is running and change LB/other stuff to point to new cluster. In the end, destroy old cluster Vms.

This would (almost?) eliminate service disruption would it not?

**Comment by a-robinson** [comment]
I suppose you could do that so long as you either
(1) store etcd's data on a persistent disk that you swap over to the new machine when it's ready (in which case cloning the current machine seems to be of questionable value versus spinning up a new one)
or
(2) run etcd in clustering mode such that the new master also receives all writes that the current master accepts during the failover process

In the first case, I don't see any clear value in cloning the VM versus just rebuilding from scratch.

The second option sounds like it may be promising so long as adding and removing nodes from etcd clusters is easy, which I don't know anything about yet. I'm also not sure what would happen to writes that occur in between when the clone happens and the new etcd instance clusters with the old one.

**Comment by pires** [comment]
@a-robinson that's exactly why I suggested https://github.com/GoogleCloudPlatform/kubernetes/issues/3476.

**Comment by pires** [comment]
@a-robinson etcd 2.0.0 is out and check this out ```etcdctl member {list|add|remove}``` ;-)

**Comment by zmerlynn** [comment]
Per our conversations at the meet-up, I opened a tighter issue for 1.0 (#4855), moved this out and de-prioritized.

**Comment by bgrant0607** [comment]
Kubelet upgrade issue is #1573

**Comment by zmerlynn** [comment]
Why were this and #1573 moved back into v1.0? I opened #4855 for the 1.0 stake in the ground (and moved this out). These bugs both have tons of items that are out of scope, and are perfectly fine discussion bugs, but don't represent the agreed 1.0 line in the sand.

(We can talk about that line in #4855, but it's nowhere near the contents of both of these bugs.)

cc @alex-mohr

**Comment by mbforbes** [comment]
With the roundup of cluster upgrades into two roll-up issues&mdash;#6075 for master, #6079 for node&mdash;I'm closing discussion issues that preceded them like this one. Those issues reference this one, and their work items are inspired from the discussions and conclusions that were drawn here.

Please feel free to re-open this issue if it contains a clear work item I missed!
